lib/launcher/test/unit/Cardano/LauncherSpec.hs

```diff
@@ -40,78 +89,155 @@ spec = do
+        (phs, ProcessHasExited name code) <- launch tr commands
         name `shouldBe` cmdName (commands !! 1)
         code `shouldBe` ExitSuccess
+        assertProcessesExited phs
 
-    it "1st process exits with 3, others are cancelled" $ do
+    it "1st process exits with 1, others are cancelled" $ \MockCommands{..} -> withTestLogging $ \tr -> do
         let commands =
               [ mockCommand False (pure ())
               , foreverCommand
               ]
-        (ProcessHasExited name code) <- launch nullTracer commands
+        (phs, ProcessHasExited name code) <- launch tr commands
         name `shouldBe` cmdName (commands !! 0)
-        code `shouldBe` (ExitFailure 3)
+        code `shouldBe` (ExitFailure 1)
+        assertProcessesExited phs
 
-    it "2nd process exits with 3, others are cancelled" $ do
+    it "2nd process exits with 1, others are cancelled" $ \MockCommands{..} -> withTestLogging $ \tr -> do
         let commands =
               [ foreverCommand
               , mockCommand False (pure ())
               ]
-        (ProcessHasExited name code) <- launch nullTracer commands
+        (phs, ProcessHasExited name code) <- launch tr commands
         name `shouldBe` cmdName (commands !! 1)
-        code `shouldBe` (ExitFailure 3)
+        code `shouldBe` (ExitFailure 1)
+        assertProcessesExited phs
 
-    it "Process executes a command before they start" $ do
+    it "Process executes a command before they start" $ \MockCommands{..} -> withTestLogging $ \tr -> do
         mvar <- newEmptyMVar
         let before = putMVar mvar "executed"
         let commands =
                 [ mockCommand True before
                 ]
-        (ProcessHasExited _ code) <- launch nullTracer commands
+        (phs, ProcessHasExited _ code) <- launch tr commands
         code `shouldBe` ExitSuccess
         tryReadMVar mvar `shouldReturn` (Just @String "executed")
+        assertProcessesExited phs
 
-    it "Handles command not found" $ do
+    it "Handles command not found" $ \MockCommands{..} -> withTestLogging $ \tr -> do
         let commands =
                 [ Command "foobar" [] (pure ()) Inherit
                 ]
-        ProcessDidNotStart name _exc <- launch nullTracer commands
+        (phs, ProcessDidNotStart name _exc) <- launch tr commands
         name `shouldBe` "foobar"
+        assertProcessesExited phs
 
-    it "Sanity check System.Info.os" $
+    it "Backend process is terminated when Async thread is cancelled" $ \MockCommands{..} -> withTestLogging $ \tr -> do
+        mvar <- newEmptyMVar
+        let backend = withBackendProcessHandle tr foreverCommand $ \ph -> do
+                putMVar mvar ph
+                forever $ threadDelay maxBound
+        before <- getCurrentTime
+        race_ backend (threadDelay 1000000)
+        after <- getCurrentTime
+        ph <- takeMVar mvar
+        assertProcessesExited [ph]
+        diffUTCTime after before `shouldSatisfy` (< 2)
```

Comment:
It is not obvious that this `2` (seconds) actually refers to `1000000` (microseconds) above with a safe margin (which is what I assume).

Reply:
True, fixed with a comment.